File: src/adapters/whatsapp_cloud.ts
// ═══════════════════════════════════════════════════
// Agent-02 — WhatsApp Business Cloud API Adapter
// Official Meta Graph API — NO web-scraping
// ═══════════════════════════════════════════════════

import { BaseAdapter } from './base.js';
import { log } from '../gateway/eventbus.js';
import * as crypto from 'crypto';

const GRAPH_API = 'https://graph.facebook.com/v21.0';

export class WhatsAppCloudAdapter extends BaseAdapter {
    readonly platform = 'whatsapp';
    private phoneNumberId: string;
    private accessToken: string;
    private appSecret: string;
    private verifyToken: string;

    constructor(config: { phoneNumberId: string; accessToken: string; appSecret: string; verifyToken: string }) {
        super();
        this.phoneNumberId = config.phoneNumberId;
        this.accessToken = config.accessToken;
        this.appSecret = config.appSecret;
        this.verifyToken = config.verifyToken;
    }

    async connect(): Promise<void> {
        this.setupOutbound();
        log('info', 'whatsapp', 'WhatsApp Business Cloud API adapter ready (webhook mode)');
        log('info', 'whatsapp', `Configure webhook URL: https://your-domain/api/webhooks/whatsapp`);
    }

    async disconnect(): Promise<void> {
        log('info', 'whatsapp', 'Adapter stopped');
    }

    // ── Webhook Verification (GET) ──
    verifyWebhook(mode: string, token: string, challenge: string): string | null {
        if (mode === 'subscribe' && token === this.verifyToken) {
            log('info', 'whatsapp', 'Webhook verified successfully');
            return challenge;
        }
        log('warn', 'whatsapp', 'Webhook verification failed');
        return null;
    }

    // ── Webhook Signature Validation ──
    validateSignature(payload: string, signature: string): boolean {
        const expected = crypto
            .createHmac('sha256', this.appSecret)
            .update(payload)
            .digest('hex');
        return `sha256=${expected}` === signature;
    }

    // ── Incoming Webhook (POST) ──
    handleWebhook(body: any): void {
        try {
            const entries = body?.entry || [];
            for (const entry of entries) {
                const changes = entry?.changes || [];
                for (const change of changes) {
                    if (change.field !== 'messages') continue;
                    const value = change.value;
                    const messages = value?.messages || [];
                    const contacts = value?.contacts || [];

                    for (const msg of messages) {
                        const from = msg.from; // phone number
                        const contact = contacts.find((c: any) => c.wa_id === from);
                        const name = contact?.profile?.name || from;

                        let content = '';
                        if (msg.type === 'text') {
                            content = msg.text?.body || '';
                        } else {
                            content = `[${msg.type} message - not supported yet]`;
                        }

                        if (content) {
                            this.emitInbound(from, name, content);
                        }
                    }
                }
            }
        } catch (err: any) {
            log('error', 'whatsapp', `Webhook parse error: ${err.message}`);
        }
    }

    // ── Send Message via Cloud API ──
    async sendMessage(to: string, content: string): Promise<void> {
        const url = `${GRAPH_API}/${this.phoneNumberId}/messages`;

        const res = await fetch(url, {
            method: 'POST',
            headers: {
                'Authorization': `Bearer ${this.accessToken}`,
                'Content-Type': 'application/json',
            },
            body: JSON.stringify({
                messaging_product: 'whatsapp',
                recipient_type: 'individual',
                to,
                type: 'text',
                text: { preview_url: false, body: content.slice(0, 4096) },
            }),
        });

        if (!res.ok) {
            const err = await res.text();
            log('error', 'whatsapp', `Send failed: ${err}`);
            throw new Error(`WhatsApp send failed: ${res.status}`);
        }
    }
}

File: src/adapters/whatsapp_cloud.go
package adapters

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
	"time"

	"github.com/gofiber/fiber/v2"
	"github.com/yourname/agent-02/src/security"
)

type WhatsAppCloudAdapter struct {
	phoneNumberID string
	accessToken   string
	appSecret     string
	verifyToken   string
	apiVersion    string
	httpClient    *http.Client
}

type waWebhookPayload struct {
	Entry []struct {
		Changes []struct {
			Value struct {
				Messages []struct {
					From string `json:"from"`
					Text struct {
						Body string `json:"body"`
					} `json:"text"`
				} `json:"messages"`
			} `json:"value"`
		} `json:"changes"`
	} `json:"entry"`
}

func NewWhatsAppCloudAdapter(phoneNumberID, accessToken, appSecret, verifyToken, apiVersion string) *WhatsAppCloudAdapter {
	if apiVersion == "" {
		apiVersion = "v23.0"
	}
	return &WhatsAppCloudAdapter{
		phoneNumberID: phoneNumberID,
		accessToken:   accessToken,
		appSecret:     appSecret,
		verifyToken:   verifyToken,
		apiVersion:    apiVersion,
		httpClient: &http.Client{
			Timeout: 20 * time.Second,
		},
	}
}

func (a *WhatsAppCloudAdapter) Enabled() bool {
	return a.phoneNumberID != "" && a.accessToken != ""
}

func (a *WhatsAppCloudAdapter) VerifyWebhook(c *fiber.Ctx) error {
	mode := c.Query("hub.mode")
	token := c.Query("hub.verify_token")
	challenge := c.Query("hub.challenge")
	if mode == "subscribe" && token == a.verifyToken {
		return c.SendString(challenge)
	}
	return fiber.ErrUnauthorized
}

func (a *WhatsAppCloudAdapter) ParseWebhook(c *fiber.Ctx) (InboundMessage, error) {
	if !security.VerifyMetaSignature(c.Body(), a.appSecret, c.Get("X-Hub-Signature-256")) {
		return InboundMessage{}, fiber.ErrUnauthorized
	}

	var payload waWebhookPayload
	if err := c.BodyParser(&payload); err != nil {
		return InboundMessage{}, err
	}

	for _, entry := range payload.Entry {
		for _, change := range entry.Changes {
			for _, msg := range change.Value.Messages {
				if strings.TrimSpace(msg.Text.Body) == "" {
					continue
				}
				return InboundMessage{
					Platform:  "whatsapp",
					ChannelID: msg.From,
					UserID:    msg.From,
					Text:      msg.Text.Body,
				}, nil
			}
		}
	}

	return InboundMessage{}, fiber.ErrNoContent
}

func (a *WhatsAppCloudAdapter) SendText(ctx context.Context, channelID string, text string) error {
	payload := map[string]any{
		"messaging_product": "whatsapp",
		"recipient_type":    "individual",
		"to":                channelID,
		"type":              "text",
		"text": map[string]any{
			"body": text,
		},
	}
	buf := &bytes.Buffer{}
	_ = json.NewEncoder(buf).Encode(payload)

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, a.messagesEndpoint(), buf)
	if err != nil {
		return err
	}
	req.Header.Set("Authorization", "Bearer "+a.accessToken)
	req.Header.Set("Content-Type", "application/json")

	resp, err := a.httpClient.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode >= 300 {
		body, _ := io.ReadAll(io.LimitReader(resp.Body, 1<<20))
		return fmt.Errorf("whatsapp send failed: %s %s", resp.Status, strings.TrimSpace(string(body)))
	}
	return nil
}

func (a *WhatsAppCloudAdapter) SelfTest(ctx context.Context) error {
	u := fmt.Sprintf("https://graph.facebook.com/%s/%s?fields=display_phone_number", a.apiVersion, url.PathEscape(a.phoneNumberID))
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, u, nil)
	if err != nil {
		return err
	}
	req.Header.Set("Authorization", "Bearer "+a.accessToken)
	resp, err := a.httpClient.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 300 {
		body, _ := io.ReadAll(io.LimitReader(resp.Body, 1<<20))
		return fmt.Errorf("whatsapp self test failed: %s %s", resp.Status, strings.TrimSpace(string(body)))
	}
	return nil
}

func (a *WhatsAppCloudAdapter) messagesEndpoint() string {
	return fmt.Sprintf("https://graph.facebook.com/%s/%s/messages", a.apiVersion, url.PathEscape(a.phoneNumberID))
}